File: arith_intensity.py
#!/usr/bin/env python3
"""Arithmetic intensity of the BK kernels -- closed-form model and an
experiment on whether the contraction order matters.

This mirrors, in a few lines of Python, what ``arith_intensity.cpp`` measures
by instrumenting the real C++ kernels. The two agree exactly (the C++ tool
asserts its measurements against these same formulas), so this file is the
convenient place to explore parameter ranges and answer "does the order of the
contractions change the flop count?".

Conventions (identical to the C++ tool):
  * nq quadrature points per direction; for BK1/BK3, nm = nq-1 modes, p = nq-2.
  * a*b counts as 2 flops (one multiply + one add); a hardware FMA is still 2.
  * bytes = streaming DRAM traffic per element = read `in` + write `out` +
    read the geometric factors. Basis matrices are cached/reused -> excluded;
    scratch is register/local memory -> excluded.
  * bytes use 4 (float), the type the kernels are instantiated with. Using
    double doubles the bytes and halves the intensity.
"""

from itertools import permutations

BYTES_PER_SCALAR = 4  # float


# --------------------------------------------------------------------------
# Closed-form flop counts (adds, muls) per element. A contraction step
# `acc = 0; for n<L: acc += a*b` over S outputs costs L*S adds and L*S muls.
# --------------------------------------------------------------------------
def bk1_counts(nq):
    m, q = nq - 1, nq
    shared = 2 * (q * m**3 + q**2 * m**2 + m * q**3)
    return shared, shared + q**3            # (adds, muls); +q^3 muls from JxW


def bk3_counts(nq):
    m, q = nq - 1, nq
    interp = 2 * (q * m**3 + q**2 * m**2 + m * q**3)
    return interp + 6 * q**4 + 6 * q**3, interp + 6 * q**4 + 9 * q**3


def bk5_counts(nq):
    q = nq
    return 6 * q**4 + 6 * q**3, 6 * q**4 + 9 * q**3


def bk1_bytes(nq):
    m, q = nq - 1, nq
    return BYTES_PER_SCALAR * (2 * m**3 + q**3)        # in + out + JxW


def bk3_bytes(nq):
    m, q = nq - 1, nq
    return BYTES_PER_SCALAR * (2 * m**3 + 6 * q**3)    # in + out + G(6)


def bk5_bytes(nq):
    q = nq
    return BYTES_PER_SCALAR * (2 * q**3 + 6 * q**3)    # in + out + G(6)


def bk1_dofs(nq):
    return (nq - 1) ** 3


bk3_dofs = bk1_dofs


def bk5_dofs(nq):
    return nq**3


# Order axis p = 1..16. BK1/BK3 over-integrate with nq = p+2; BK5 collocates
# at the p+1 GLL nodes, so nq = p+1.
KERNELS = {
    "BK1": (bk1_counts, bk1_bytes, bk1_dofs, range(3, 19), 2),  # nq = p+2
    "BK3": (bk3_counts, bk3_bytes, bk3_dofs, range(3, 19), 2),  # nq = p+2
    "BK5": (bk5_counts, bk5_bytes, bk5_dofs, range(2, 18), 1),  # nq = p+1
}


def print_tables():
    for name, (counts, nbytes, ndofs, nq_range, p_off) in KERNELS.items():
        print(f"\n{name}  (nq = p+{p_off})")
        print(f"{'p':>4}{'nq':>4}{'flops':>12}{'bytes':>9}"
              f"{'flop/DoF':>11}{'flop/byte':>12}{'byte/flop':>12}")
        for nq in nq_range:
            a, mul = counts(nq)
            flops = a + mul
            b = nbytes(nq)
            d = ndofs(nq)
            print(f"{nq - p_off:>4}{nq:>4}{flops:>12}{b:>9}"
                  f"{flops / d:>11.3f}{flops / b:>12.3f}{b / flops:>12.3f}")


# --------------------------------------------------------------------------
# Does the contraction order matter?
#
# The forward interpolation contracts one Cartesian direction at a time,
# turning that direction's `m` modes into `q` points. When direction d is
# contracted, the directions already contracted contribute a factor q each and
# those not yet contracted contribute m each; the reduction length is m[d]:
#
#     cost(step) = 2 * m[d] * q[d] * prod(q over done) * prod(m over pending)
#
# so the total over a permutation depends on the intermediate tensor sizes,
# which depend on the order -- unless every direction has the same (m, q).
# --------------------------------------------------------------------------
def forward_interp_flops(m, q, order):
    total, done = 0, set()
    for d in order:
        out = q[d]
        for a in range(3):
            if a == d:
                continue
            out *= q[a] if a in done else m[a]
        total += 2 * m[d] * out       # reduction length m[d], counted as 2 flops
        done.add(d)
    return total


def order_experiment():
    print("\nDoes the contraction order matter? (forward interpolation)\n"
          "Total flops for each of the 6 direction orderings:")
    cases = {
        "isotropic   m=(4,4,4) q=(5,5,5)": ([4, 4, 4], [5, 5, 5]),
        "anisotropic m=(2,4,6) q=(3,5,7)": ([2, 4, 6], [3, 5, 7]),
    }
    for label, (m, q) in cases.items():
        vals = {order: forward_interp_flops(m, q, order)
                for order in permutations(range(3))}
        lo, hi = min(vals.values()), max(vals.values())
        print(f"\n  {label}")
        for order, v in vals.items():
            print(f"    order {order} -> {v:>7} flops")
        if lo == hi:
            print(f"    => all identical ({lo}); order is irrelevant.")
        else:
            best = min(vals, key=vals.get)
            print(f"    => range {lo}..{hi} "
                  f"({100*(hi-lo)/lo:.1f}% spread); best order {best}.")


if __name__ == "__main__":
    print("Per-element arithmetic intensity of the BK kernels "
          f"(bytes: float = {BYTES_PER_SCALAR}; a*b = 2 flops)")
    print_tables()
    order_experiment()

File: arith_intensity.cpp
// ---------------------------------------------------------------------------
// arith_intensity: exact flop / byte / arithmetic-intensity report for the
// BK kernels.
//
// Two independent methods are combined and cross-checked:
//
//   1. Measurement. Each kernel is instantiated with flop_counter<double> in
//      place of the value type and run once (nelmt = 1). Every arithmetic
//      operator the *actual kernel source* performs bumps a counter, so the
//      reported flops are exact -- there is no modelling of the loop bodies.
//
//   2. Closed form. Analytic flop formulas (derived per step, see below) are
//      evaluated and asserted equal to the measured counts. If they ever
//      disagree the program exits non-zero, so this doubles as a regression
//      test on the formulas.
//
// Bytes counted are the streaming DRAM traffic per element, matching the
// "effective GB/s" the timing drivers report: read `in`, write `out`, read
// the per-point geometric factors (JxW for BK1, the 6-component G for BK3 /
// BK5). The small basis / derivative matrices are reused across all elements
// and cached, so -- like the drivers -- they are excluded. Scratch space is
// register / local memory, not DRAM traffic, and is likewise excluded.
//
// Arithmetic intensity = flops / bytes. Bytes use sizeof(float) = 4, the type
// the kernels are actually instantiated with; using double doubles the bytes
// and so halves the intensity (flops are unchanged).
// ---------------------------------------------------------------------------

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <utility>
#include <vector>

#include "flop_counter.h"
#include "bk1_kernel.h"
#include "bk3_kernel.h"
#include "bk5_kernel.h"

namespace {

using Real = double;             // wrapped value type -- values are irrelevant
using C    = bk::flop_counter<Real>;

// Bytes per scalar in the deployed kernels (float). Change to 8 for double.
constexpr long long BYTES_PER_SCALAR = sizeof(float);

bool g_ok = true;               // cleared on any formula/measurement mismatch

struct Counts { long long adds, muls; };

constexpr long long ipow(long long b, int e)
{
    long long r = 1;
    for (int i = 0; i < e; ++i) r *= b;
    return r;
}

// -------- closed-form flop formulas (adds, muls) per element ----------------
// A contraction step `acc = 0; for n<L: acc += a*b` over S outputs contributes
// L*S adds and L*S muls. Summing the per-step contributions:

constexpr Counts formula_bk1(int nq)
{
    const long long m = nq - 1, q = nq;
    const long long shared = 2*q*ipow(m,3) + 2*q*q*m*m + 2*m*ipow(q,3);
    return { shared,                 // adds
             shared + ipow(q,3) };   // muls: extra nq^3 from the JxW scaling
}

constexpr Counts formula_bk3(int nq)
{
    const long long m = nq - 1, q = nq;
    const long long interp = 2*q*ipow(m,3) + 2*q*q*m*m + 2*m*ipow(q,3);
    return { interp + 6*ipow(q,4) + 6*ipow(q,3),    // adds
             interp + 6*ipow(q,4) + 9*ipow(q,3) };  // muls
}

constexpr Counts formula_bk5(int nq)
{
    const long long q = nq;
    return { 6*ipow(q,4) + 6*ipow(q,3),      // adds
             6*ipow(q,4) + 9*ipow(q,3) };    // muls
}

// ----------------------------- measurement ----------------------------------
// Fill helper: values do not affect the (data-independent) trip counts.
std::vector<C> box(std::size_t n) { return std::vector<C>(n, C(Real(1))); }

template <int nq>
Counts measure_bk1()
{
    constexpr int nm = nq - 1;
    auto basis = box(nm*nq), JxW = box(nq*nq*nq),
         in    = box(nm*nm*nm), out = box(nm*nm*nm);
    C::reset();
    bk::bk1::SumFactorization<C, nq>(1, basis.data(), JxW.data(),
                                     in.data(), out.data());
    return { (long long)C::adds, (long long)C::muls };
}

template <int nq>
Counts measure_bk3()
{
    constexpr int nm = nq - 1;
    auto basis = box(nm*nq), dbasis = box(nq*nq), G = box(6*nq*nq*nq),
         in    = box(nm*nm*nm), out = box(nm*nm*nm);
    C::reset();
    bk::bk3::SumFactorization<C, nq>(1, basis.data(), dbasis.data(),
                                     G.data(), in.data(), out.data());
    return { (long long)C::adds, (long long)C::muls };
}

template <int nq>
Counts measure_bk5()
{
    auto dbasis = box(nq*nq), G = box(6*nq*nq*nq),
         in     = box(nq*nq*nq), out = box(nq*nq*nq);
    C::reset();
    bk::bk5::SumFactorization<C, nq>(1, dbasis.data(), G.data(),
                                     in.data(), out.data());
    return { (long long)C::adds, (long long)C::muls };
}

// ------------------------------- reporting ----------------------------------
void print_header()
{
    std::cout << std::right
              << std::setw(4)  << "nq"    << std::setw(4)  << "p"
              << std::setw(12) << "adds"  << std::setw(12) << "muls"
              << std::setw(12) << "flops" << std::setw(9)  << "bytes"
              << std::setw(11) << "flop/DoF"
              << std::setw(12) << "flop/byte"
              << std::setw(12) << "byte/flop" << "\n";
}

void print_row(int nq, int p, Counts meas, Counts pred,
               long long bytes, long long dofs)
{
    if (meas.adds != pred.adds || meas.muls != pred.muls) {
        g_ok = false;
        std::cout << "  [MISMATCH nq=" << nq
                  << " measured(add=" << meas.adds << ",mul=" << meas.muls
                  << ") formula(add=" << pred.adds << ",mul=" << pred.muls
                  << ")]\n";
    }
    const long long flops = meas.adds + meas.muls;
    std::cout << std::right << std::fixed << std::setprecision(3)
              << std::setw(4)  << nq
              << std::setw(4)  << (p >= 0 ? std::to_string(p) : std::string("-"))
              << std::setw(12) << meas.adds
              << std::setw(12) << meas.muls
              << std::setw(12) << flops
              << std::setw(9)  << bytes
              << std::setw(11) << double(flops) / double(dofs)
              << std::setw(12) << double(flops) / double(bytes)
              << std::setw(12) << double(bytes) / double(flops) << "\n";
}

template <typename F, int... Ns>
void for_orders(F f, std::integer_sequence<int, Ns...>)
{
    (f(std::integral_constant<int, Ns>{}), ...);
}

} // namespace

int main()
{
    std::cout << "Per-element arithmetic intensity of the BK kernels\n"
              << "(bytes = streaming DRAM traffic, sizeof(float) = "
              << BYTES_PER_SCALAR << "; a*b counted as 2 flops)\n";

    std::cout << "\nBK1  (mass matrix; DoF = nm^3, bytes = 2*nm^3 + nq^3)\n";
    print_header();
    for_orders([](auto ic) {
        constexpr int nq = ic.value, nm = nq - 1;
        const Counts meas = measure_bk1<nq>();
        const long long bytes = BYTES_PER_SCALAR * (2*ipow(nm,3) + ipow(nq,3));
        print_row(nq, nq - 2, meas, formula_bk1(nq), bytes, ipow(nm,3));
    }, std::integer_sequence<int, 3,4,5,6,7,8,9,10>{});

    std::cout << "\nBK3  (Poisson; DoF = nm^3, bytes = 2*nm^3 + 6*nq^3)\n";
    print_header();
    for_orders([](auto ic) {
        constexpr int nq = ic.value, nm = nq - 1;
        const Counts meas = measure_bk3<nq>();
        const long long bytes = BYTES_PER_SCALAR * (2*ipow(nm,3) + 6*ipow(nq,3));
        print_row(nq, nq - 2, meas, formula_bk3(nq), bytes, ipow(nm,3));
    }, std::integer_sequence<int, 3,4,5,6,7,8,9,10>{});

    std::cout << "\nBK5  (collocated Laplacian; DoF = nq^3, bytes = 8*nq^3)\n";
    print_header();
    for_orders([](auto ic) {
        constexpr int nq = ic.value;
        const Counts meas = measure_bk5<nq>();
        const long long bytes = BYTES_PER_SCALAR * (8*ipow(nq,3));
        print_row(nq, -1, meas, formula_bk5(nq), bytes, ipow(nq,3));
    }, std::integer_sequence<int, 2,3,4,5,6,7,8>{});

    std::cout << "\n" << (g_ok
        ? "All measured counts match the closed-form formulas.\n"
        : "WARNING: measured counts disagree with the formulas (see above).\n");
    return g_ok ? 0 : 1;
}
